docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,694 @@
+* Applying policies with and without egress
+
+## Considered Options
+
+* Creating MeshExternalService resource and MeshPassthrough policy
+* Creating MeshExternalService
+* Creating MeshExternalService + HostnameGenerator
+* Creating MeshExternalService and using HostnameGenerator for InternalVIP
+
+## Decision Outcome
+
+Chosen option: "Creating MeshExternalService resource and MeshPassthrough policy".
+
+### Positive Consequences
+
+* Clearer structure.
+* Explicit showing what is passthough and what is external service
+* Enhanced functionalities.
+* Better policy matching capabilities.
+
+### Negative Consequences
+
+* 2 new resources
+* Temporarily increased complexity of a product until the migration is done.
+
+## Pros and Cons of the Options
+
+### Creating MeshExternalService
+
+We're introducing a new object resource `MeshExternalService`. This object will be a namespace resource and will only be creatable within the `kuma-system` namespace. It can be created within a specific zone, but in this case, it will only be accessible within the zone where it was created.
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  type: Managed # Managed
+  managed:
+    match:
+      type: KumaHostname # Kuma will generate a domain
+      port: 80
+      protocol: http
+    extension:
+      type: Lambda 
+      config: # type JSON
+        arn: arn:aws:lambda:us-west-2:123456789012:function:my-function
+    endpoints:
+      - address: 1.1.1.1
+        port: 12345
+      - address: example.com
+      - address: unix://....
+    tls:
+      version:
+        min: TLSv1_2 # or TLS_AUTO, TLSv1_0, TLSv1_1, TLSv1_2, TLSv1_3
+        max: TLSv1_3 # or TLS_AUTO, TLSv1_0, TLSv1_1, TLSv1_2, TLSv1_3
+      allowRenegotiation: false
+      verification:
+        skipSAN: true # if this is true then subjectAltNames don't take effect
+        subjectAltNames: # if subjectAltNames is not defined then take domain or ips
+          - example.com
+          - "spiffe://example.local/ns/local"
+        caCert: 
+          inline: 123
+        clientCert:
+          secret: 123
+        clientKey:
+          secret: 123
+status: 
+  vip:
+    value: 242.0.0.1
+    type: Kuma
+  addresses:
+  - hostname: example.ext.svc.local
+    status: NotAvailable
+    origin:
+      kind: HostnameGenerator
+      name: k8s-example-service-hostname
+    reason: "addresses are overlapping with ext2"
+```
+* **spec**:
+  * **match**: defines a traffic that should be routed through the sidecar
+    * **type**: type of the match, one of `KumaHostname`, `Domain`, `CIDR` and `IP` are available
+      * `KumaHostname`: matches hostnames provided by HostnameGenerator
+      * `Domain`: handles traffic to the specified domain
+      * `CIDR`: handles traffic to specified addresses range
+      * `IP` handles the traffic to specified IP
```

Comment:
Are these now invalid? I thought Domain, CIDR, IP were only for passthrough which is now its own policy.

Reply:
I think you're right, they're here https://github.com/kumahq/kuma/pull/10253/files#diff-ea7b27398fed9e782b370bf3ff3d36e1d5f5cf8a2b26ef710b2feb584b717e7fR64 . Theoretically we could have one for `IP` but does it make sense? 🤔

Reply:
Yeah I don't think it does, if you want to keep the traffic the same ip just use `MeshPassthrough`

Reply:
If we want to support a case (that we support now in ESv1) that you can have an app talking to something in HTTP and then transparently to the app upgrade this to HTTPS we would use `Domain`. This might be useful when you have very old service that you don't know/want/have a way to modify but your security team requires HTTPS on the egress.

Reply:
Let's start with just `HostnameGenerator` and we can expand if needed.